File: luda-ts/editor/client/src/cameraAngleEditor/wysiwygEditor/Resizer.ts
import {
  ColorUtil,
  RenderingTree,
  Translate,
  Vector,
  PathDrawCommand,
  RenderingData,
  Cursor,
  engine,
  Rect,
  AfterDraw,
  Convert,
  Render,
  WhSize,
  XywhRect,
  LtrbRect,
  BorderPosition,
} from "namui";
import { CameraAngleEditorState, ImageSource } from "../type";
import { getSourceRect } from "./getRect";
import { update01Rect } from "./update01Rect";

export const Resizer: Render<
  CameraAngleEditorState,
  {
    containerSize: WhSize;
    imageSource: ImageSource;
  }
> = (state, props) => {
  const sourceRect: XywhRect = {
    x: props.containerSize.width * state.cameraAngle.source01Rect.x,
    y: props.containerSize.height * state.cameraAngle.source01Rect.y,
    width: props.containerSize.width * state.cameraAngle.source01Rect.width,
    height: props.containerSize.height * state.cameraAngle.source01Rect.height,
  };
  return [
    Rect({
      ...sourceRect,
      style: {
        stroke: {
          color: ColorUtil.Grayscale01(0.2),
          width: 1,
          borderPosition: BorderPosition.inside,
        },
      },
    }),
    Translate(
      {
        ...sourceRect,
      },
      [
        ImageSizeHandles(state, {
          imageSource: props.imageSource,
        }),
      ],
    ),
  ];
};
export const ImageSizeHandles: Render<
  CameraAngleEditorState,
  {
    imageSource: ImageSource;
  }
> = (state, props) => {
  const {
    imageSource: { widthHeightRatio },
  } = props;

  const handleRadius = 5;
  const handles = getHandles(state);

  return [
    ...handles.map((handle) => {
      const left = handle.position.x - handleRadius;
      const top = handle.position.y - handleRadius;
      const right = handle.position.x + handleRadius;
      const bottom = handle.position.y + handleRadius;

      const path = new CanvasKit.Path();
      path.addOval([left, top, right, bottom]);

      const fillPaint = new CanvasKit.Paint();
      fillPaint.setStyle(CanvasKit.PaintStyle.Fill);
      fillPaint.setColor(ColorUtil.White);

      const strokePaint = new CanvasKit.Paint();
      strokePaint.setStyle(CanvasKit.PaintStyle.Stroke);
      strokePaint.setColor(ColorUtil.Grayscale01(0.5));
      strokePaint.setStrokeWidth(2);
      strokePaint.setAntiAlias(true);

      return RenderingData({
        drawCalls: [
          {
            commands: [
              PathDrawCommand({
                path,
                paint: fillPaint,
              }),
              PathDrawCommand({
                path,
                paint: strokePaint,
              }),
            ],
          },
        ],
        onMouseIn() {
          engine.mousePointer.setCursor(handle.cursor);
        },
        onMouseDown(event) {
          state.wysiwygEditor.dragging = {
            targetId: handle.id,
            lastMousePosition: Vector.from(event),
          };
        },
      });
    }),
    AfterDraw(({ translated }) => {
      const sourceRect = getSourceRect(state);
      const container = new Vector(
        translated.x - sourceRect.x,
        translated.y - sourceRect.y,
      );

      engine.mouseEvent.onMouseMove((event) => {
        const { dragging } = state.wysiwygEditor;
        if (!dragging || !dragging.targetId.startsWith("resize-")) {
          return;
        }
        const sourceRect = getSourceRect(state);
        const { targetId } = dragging;
        const mouseVector = Vector.from(event);
        const diff = mouseVector.sub(dragging.lastMousePosition);

        const mouseLocalVector = mouseVector.sub(container);
        const handleId = targetId.substring("resize-".length);
        const isDiagonal = handleId.includes("-");

        if (isDiagonal) {
          const centerPoint = new Vector(
            sourceRect.x + sourceRect.width / 2,
            sourceRect.y + sourceRect.height / 2,
          );
          const endPoint = new Vector(
            handleId.includes("left")
              ? sourceRect.x
              : sourceRect.x + sourceRect.width,
            handleId.includes("top")
              ? sourceRect.y
              : sourceRect.y + sourceRect.height,
          );

          const diagonalVector = endPoint.sub(centerPoint);
          const centerToMouseLocalVector = mouseLocalVector.sub(centerPoint);

          const isCcw = diagonalVector.cross(centerToMouseLocalVector) > 0;
          function getMain(): keyof LtrbRect {
            switch (handleId) {
              case "top-left":
                return isCcw ? "top" : "left";
              case "top-right":
                return isCcw ? "right" : "top";
              case "bottom-right":
                return isCcw ? "bottom" : "right";
              case "bottom-left":
                return isCcw ? "left" : "bottom";
              default:
                throw new Error(`unreachable - ${handleId}`);
            }
          }
          const main = getMain();
          const sub = handleId
            .replace(main, "")
            .replace("-", "") as keyof LtrbRect;

          const nextLtrbRect = resizeRect({
            rect: Convert.xywhToLtrb(sourceRect),
            main,
            sub,
            mouseLocalVector,
            widthHeightRatio,
          });
          const nextRect = Convert.ltrbToXywh(nextLtrbRect);
          update01Rect(state, state.cameraAngle.source01Rect, nextRect);
        } else {
          const nextRect: XywhRect = { ...sourceRect };
          if (handleId === "top" || handleId === "bottom") {
            let deltaWidth: number;

            if (handleId === "top") {
              nextRect.y += diff.y;
              nextRect.height -= diff.y;
              deltaWidth = -diff.y * widthHeightRatio;
            } else {
              nextRect.height += diff.y;
              deltaWidth = diff.y * widthHeightRatio;
            }

            nextRect.x -= deltaWidth / 2;
            nextRect.width += deltaWidth;
          } else {
            let deltaHeight: number;

            if (handleId === "left") {
              nextRect.x += diff.x;
              nextRect.width -= diff.x;
              deltaHeight = -diff.x / widthHeightRatio;
            } else {
              nextRect.width += diff.x;
              deltaHeight = diff.x / widthHeightRatio;
            }

            nextRect.y -= deltaHeight / 2;
            nextRect.height += deltaHeight;
          }
          update01Rect(state, state.cameraAngle.source01Rect, nextRect);
        }
        dragging.lastMousePosition = mouseVector;
      });
    }),
  ];
};

function getHandles(state: CameraAngleEditorState): {
  id:
    | "resize-top-left"
    | "resize-top"
    | "resize-top-right"
    | "resize-left"
    | "resize-right"
    | "resize-bottom-left"
    | "resize-bottom"
    | "resize-bottom-right";
  position: Vector;
  cursor: Cursor;
}[] {
  const sourceRect: XywhRect = getSourceRect(state);
  return [
    {
      id: "resize-top-left",
      position: new Vector(0, 0),
      cursor: Cursor.leftTopRightBottomResize,
    },
    {
      id: "resize-top",
      position: new Vector(sourceRect.width / 2, 0),
      cursor: Cursor.topBottomResize,
    },
    {
      id: "resize-top-right",
      position: new Vector(sourceRect.width, 0),
      cursor: Cursor.rightTopLeftBottomResize,
    },
    {
      id: "resize-left",
      position: new Vector(0, sourceRect.height / 2),
      cursor: Cursor.leftRightResize,
    },
    {
      id: "resize-right",
      position: new Vector(sourceRect.width, sourceRect.height / 2),
      cursor: Cursor.leftRightResize,
    },
    {
      id: "resize-bottom-left",
      position: new Vector(0, sourceRect.height),
      cursor: Cursor.rightTopLeftBottomResize,
    },
    {
      id: "resize-bottom",
      position: new Vector(sourceRect.width / 2, sourceRect.height),
      cursor: Cursor.topBottomResize,
    },
    {
      id: "resize-bottom-right",
      position: new Vector(sourceRect.width, sourceRect.height),
      cursor: Cursor.leftTopRightBottomResize,
    },
  ];
}

function resizeRect({
  rect,
  main,
  sub,
  mouseLocalVector,
  widthHeightRatio,
}: {
  rect: LtrbRect;
  main: keyof LtrbRect;
  sub: keyof LtrbRect;
  mouseLocalVector: Vector;
  widthHeightRatio: number;
}): LtrbRect {
  const nextRect: LtrbRect = {
    ...rect,
  };

  const isMainHorizontal = main === "left" || main === "right";

  nextRect[main] = isMainHorizontal ? mouseLocalVector.x : mouseLocalVector.y;
  const deltaMain = nextRect[main] - rect[main];

  const shouldDeltaReverse =
    (main === "right" && sub === "top") ||
    (main === "top" && sub === "right") ||
    (main === "left" && sub === "bottom") ||
    (main === "bottom" && sub === "left");

  nextRect[sub] +=
    deltaMain *
    (isMainHorizontal ? 1 / widthHeightRatio : widthHeightRatio) *
    (shouldDeltaReverse ? -1 : 1);

  return nextRect;
}

File: luda-editor/luda-editor-client/src/editor/clip_editor/camera_clip_editor/wysiwyg_editor/resizer.rs
use crate::editor::types::*;
use crate::editor::EditorEvent::WysiwygEditorResizerHandleMouseDownEvent;
use namui::prelude::*;

pub struct Resizer {}

impl Resizer {
    pub fn new() -> Self {
        Self {}
    }
}

pub struct ResizerProps<'a> {
    pub camera_angle: &'a CameraAngle,
    pub source_rect: &'a XywhRect<f32>,
    pub container_size: &'a Wh<f32>,
}

impl Resizer {
    pub fn update(&mut self, event: &dyn std::any::Any) {}

    pub fn render(&self, props: &ResizerProps) -> RenderingTree {
        render![
            rect(RectParam {
                x: props.source_rect.x,
                y: props.source_rect.y,
                width: props.source_rect.width,
                height: props.source_rect.height,
                style: RectStyle {
                    stroke: Some(RectStroke {
                        color: Color::gary_scale_f01(0.2),
                        width: 1.0,
                        border_position: BorderPosition::Inside,
                    }),
                    ..Default::default()
                },
            }),
            render_resize_handles(props.source_rect, props.container_size),
        ]
    }
}

fn render_resize_handles(source_rect: &XywhRect<f32>, container_size: &Wh<f32>) -> RenderingTree {
    const HANDLE_RADIUS: f32 = 5.0;

    RenderingTree::Children(
        get_handles(&source_rect)
            .iter()
            .map(|handle| {
                let path = namui::Path::new().add_oval(&LtrbRect {
                    left: handle.xy.x - HANDLE_RADIUS,
                    top: handle.xy.y - HANDLE_RADIUS,
                    right: handle.xy.x + HANDLE_RADIUS,
                    bottom: handle.xy.y + HANDLE_RADIUS,
                });

                let fill_paint = namui::Paint::new()
                    .set_style(namui::PaintStyle::Fill)
                    .set_color(Color::WHITE);

                let stroke_paint = namui::Paint::new()
                    .set_style(namui::PaintStyle::Stroke)
                    .set_color(Color::gary_scale_f01(0.5))
                    .set_stroke_width(2.0)
                    .set_anti_alias(true);

                render![
                    namui::path(path.clone(), fill_paint),
                    namui::path(path, stroke_paint),
                ]
                .with_mouse_cursor(handle.cursor())
                .attach_event(move |builder| {
                    let handle = handle.clone();
                    let container_size = container_size.clone();
                    let source_rect = source_rect.clone();
                    builder.on_mouse_down(Box::new(move |mouse_event| {
                        namui::event::send(Box::new(WysiwygEditorResizerHandleMouseDownEvent {
                            handle,
                            center_xy: source_rect.center(),
                            mouse_xy: mouse_event.global_xy,
                            container_size,
                            image_size_ratio: Wh {
                                width: source_rect.width,
                                height: source_rect.height,
                            },
                        }))
                    }))
                })
            })
            .collect::<Vec<RenderingTree>>(),
    )
}

fn get_opposite_handle(handle: &ResizerHandle, source_rect: &XywhRect<f32>) -> ResizerHandle {
    let center_xy = source_rect.center();
    ResizerHandle {
        handle_direction: handle.handle_direction.opposite(),
        xy: 2.0 * center_xy - handle.xy,
    }
}

#[derive(Debug, Clone, Copy)]
pub enum ResizerHandleDirection {
    TopLeft,
    TopRight,
    BottomRight,
    BottomLeft,
    Top,
    Right,
    Bottom,
    Left,
}
#[derive(Debug, Clone, Copy)]
pub struct ResizerHandle {
    pub handle_direction: ResizerHandleDirection,
    pub xy: Xy<f32>,
}
impl ResizerHandle {
    pub fn cursor(&self) -> namui::MouseCursor {
        self.handle_direction.cursor()
    }
}
fn get_handles(source_rect: &XywhRect<f32>) -> Vec<ResizerHandle> {
    let top_left = ResizerHandle {
        handle_direction: ResizerHandleDirection::TopLeft,
        xy: Xy {
            x: source_rect.x,
            y: source_rect.y,
        },
    };

    let top = ResizerHandle {
        handle_direction: ResizerHandleDirection::Top,
        xy: Xy {
            x: source_rect.x + source_rect.width / 2.0,
            y: source_rect.y,
        },
    };
    let top_right = ResizerHandle {
        handle_direction: ResizerHandleDirection::TopRight,
        xy: Xy {
            x: source_rect.x + source_rect.width,
            y: source_rect.y,
        },
    };
    let left = ResizerHandle {
        handle_direction: ResizerHandleDirection::Left,
        xy: Xy {
            x: source_rect.x,
            y: source_rect.y + source_rect.height / 2.0,
        },
    };
    vec![
        top_left,
        top,
        top_right,
        left,
        get_opposite_handle(&left, &source_rect),
        get_opposite_handle(&top_right, &source_rect),
        get_opposite_handle(&top, &source_rect),
        get_opposite_handle(&top_left, &source_rect),
    ]
}
impl ResizerHandleDirection {
    pub(crate) fn opposite(&self) -> Self {
        match self {
            ResizerHandleDirection::TopLeft => ResizerHandleDirection::BottomRight,
            ResizerHandleDirection::TopRight => ResizerHandleDirection::BottomLeft,
            ResizerHandleDirection::BottomLeft => ResizerHandleDirection::TopRight,
            ResizerHandleDirection::BottomRight => ResizerHandleDirection::TopLeft,
            ResizerHandleDirection::Top => ResizerHandleDirection::Bottom,
            ResizerHandleDirection::Bottom => ResizerHandleDirection::Top,
            ResizerHandleDirection::Left => ResizerHandleDirection::Right,
            ResizerHandleDirection::Right => ResizerHandleDirection::Left,
        }
    }
    pub(crate) fn cursor(&self) -> MouseCursor {
        match self {
            ResizerHandleDirection::TopLeft => MouseCursor::LeftTopRightBottomResize,
            ResizerHandleDirection::Top => MouseCursor::TopBottomResize,
            ResizerHandleDirection::TopRight => MouseCursor::RightTopLeftBottomResize,
            ResizerHandleDirection::Left => MouseCursor::LeftRightResize,
            ResizerHandleDirection::Right => MouseCursor::LeftRightResize,
            ResizerHandleDirection::BottomLeft => MouseCursor::RightTopLeftBottomResize,
            ResizerHandleDirection::Bottom => MouseCursor::TopBottomResize,
            ResizerHandleDirection::BottomRight => MouseCursor::LeftTopRightBottomResize,
        }
    }
}
